src/backend/distributed/transaction/remote_transaction.c

```diff
@@ -820,6 +845,247 @@ CoordinatedRemoteTransactionsAbort(void)
 
 
 /*
+ * RemoteTransactionsSavepointBegin sends the SAVEPOINT command for the given
+ * sub-transaction id to all connections participating in the current transaction.
+ */
+void
+RemoteTransactionsSavepointBegin(SubTransactionId subId)
+{
+	dlist_iter iter;
+
+	/* asynchronously send SAVEPOINT */
+	dlist_foreach(iter, &InProgressTransactions)
+	{
+		MultiConnection *connection = dlist_container(MultiConnection, transactionNode,
+													  iter.cur);
+		RemoteTransaction *transaction = &connection->remoteTransaction;
+		if (transaction->transactionFailed)
+		{
+			continue;
+		}
+
+		StartRemoteTransactionSavepointBegin(connection, subId);
+	}
+
+	/* and wait for the results */
+	dlist_foreach(iter, &InProgressTransactions)
+	{
+		MultiConnection *connection = dlist_container(MultiConnection, transactionNode,
+													  iter.cur);
+		RemoteTransaction *transaction = &connection->remoteTransaction;
+
+		if (transaction->transactionFailed)
+		{
+			continue;
+		}
+
+		FinishRemoteTransactionSavepointBegin(connection, subId);
+
+		if (!transaction->transactionFailed)
+		{
+			transaction->lastUnfailedSubXact = subId;
+		}
+	}
+}
+
+
+/*
+ * RemoteTransactionsSavepointRelease sends the RELEASE SAVEPOINT command for the
+ * given sub-transaction id to all connections participating in the current transaction.
+ */
+void
+RemoteTransactionsSavepointRelease(SubTransactionId subId)
+{
+	dlist_iter iter;
+
+	/* asynchronously send SAVEPOINT */
+	dlist_foreach(iter, &InProgressTransactions)
+	{
+		MultiConnection *connection = dlist_container(MultiConnection, transactionNode,
+													  iter.cur);
+		RemoteTransaction *transaction = &connection->remoteTransaction;
+		if (transaction->transactionFailed)
+		{
+			continue;
+		}
+
+		StartRemoteTransactionSavepointRelease(connection, subId);
+	}
+
+	/* and wait for the results */
+	dlist_foreach(iter, &InProgressTransactions)
+	{
+		MultiConnection *connection = dlist_container(MultiConnection, transactionNode,
+													  iter.cur);
+		RemoteTransaction *transaction = &connection->remoteTransaction;
+
+		if (transaction->transactionFailed)
+		{
+			continue;
+		}
+
+		FinishRemoteTransactionSavepointRelease(connection, subId);
+	}
+}
+
+
+/*
+ * RemoteTransactionsSavepointRollback sends the ROLLBACK TO SAVEPOINT command for the
+ * given sub-transaction id to all connections participating in the current transaction.
+ */
+void
+RemoteTransactionsSavepointRollback(SubTransactionId subId)
+{
+	dlist_iter iter;
+
+	/* asynchronously send SAVEPOINT */
+	dlist_foreach(iter, &InProgressTransactions)
+	{
+		MultiConnection *connection = dlist_container(MultiConnection, transactionNode,
+													  iter.cur);
+		RemoteTransaction *transaction = &connection->remoteTransaction;
+		if (transaction->lastUnfailedSubXact <= subId &&
```

Comment:
The function header needs to clearly explain this. Also, should we really reset it before having succeeded sending the rollback?